### PR TITLE
Cutting counter for individual keys

### PR DESCRIPTION
**Pull request by cernoch**
Number of individual keys can be counted using sharpSAT, interfaced via `mks::cutcount`.

The build only works on a) modern Linux, b) Travis with clang, because:
- Windows is not supported by sharpSAT.
- Travis uses GCC 4.9, which is [known not to work with old GMP](https://github.com/evaleev/libint/issues/29).

Both issues should be resolved by porting sharpSAT to Windows (and switching to MPIR instead of GMP).

### CODE REVIEW

**Comment by cernoch** [comment]
I guess we should **postpone the merge** until the sharpSAT library is ported to Windows...

**Comment by horenmar** [comment]
Agreed.

This shouldn't take long now, as merging master-keying/sharpSAT#7 will provide Windows support.

**Comment by horenmar** [comment]
How did the `assert`s in sharpSAT end up? Because the cutcount tests on Travis report this
```
tests: /home/travis/build/master-keying/mks/lib/sharpSAT/include/sharpSAT/component_types/difference_packed_component.h:118: sharpSAT::DifferencePackedComponent::DifferencePackedComponent(sharpSAT::Component &): Assertion `(data_size >> bits_of_data_size()) == 0' failed.
/home/travis/build/master-keying/mks/src/test/cpp/cutcount_test.cpp:378: failed: fatal error condition with message: 'SIGABRT - Abort (abnormal termination) signal'
Failed 1 test case, failed 1 assertion.
```

**Comment by horenmar** [comment]
Ok, if we solve the assertion mentioned in my previous comment, this pull request works.

I will need to do some work on the coverage helper though, as it seems to hang indefinitely when the runee is aborted.

File: src/main/cpp/flatset.h [hunk 1]
```diff
@@ -168,6 +168,159 @@ class flatset {
         return data != rhs.data;
     }
 
+    /**
+     * Retain only only those elements found in rhs.
```
[the remark this answers is not in the document]
Comment (cernoch) [reply src/main/cpp/flatset.h]: Fixed.

File: src/main/cpp/flatset.h [hunk 2]
```diff
@@ -168,6 +168,159 @@ class flatset {
...
+            }
+        }
+
+        // move remaining items if something was deleted
+        data.erase(destination, std::end(data));
+        return *this;
+    }
+
+    /**
+     * Retain only only those elements found in rhs.
```
[the remark this answers is not in the document]
Comment (cernoch) [reply src/main/cpp/flatset.h]: Fixed.

File: src/main/cpp/flatset.h [hunk 3]
```diff
@@ -168,6 +168,159 @@ class flatset {
...
+        std::set_intersection(
+            std::begin(data), std::end(data),
+            std::begin(rhs.data), std::end(rhs.data),
+            std::back_inserter(items) );
+
+        return flatset<T>(std::move(items), true);
+    }
+
+    /**
+     * Retain only only those elements found in rhs.
```
[the remark this answers is not in the document]
Comment (cernoch) [reply src/main/cpp/flatset.h]: Fixed.

File: src/main/cpp/flatset.h [hunk 4]
```diff
@@ -168,6 +168,159 @@ class flatset {
...
+                ++candidate;
+            }
+        }
+
+        if (candidate != destination) { // if something was deleted
+            while (candidate != std::end(data)) { // move remaining items
+                *destination = std::move(*candidate);
+                ++candidate;
+                ++destination;
+            }
```
[the remark this answers is not in the document]
Comment (cernoch) [reply src/main/cpp/flatset.h]: Thanks for letting me know. Didn't know about this. Implementation has been updated.

File: src/main/cpp/flatset.h [hunk 5]
```diff
@@ -168,6 +168,159 @@ class flatset {
...
+            data.erase(destination, std::end(data));
+        }
+        return *this;
+    }
+
+    /**
+     * Retain only only those elements found in rhs.
```
[the remark this answers is not in the document]
Comment (cernoch) [reply src/main/cpp/flatset.h]: Fixed.

File: src/main/cpp/cutcount.h [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+     * Export the CNF instance to the DIMACS format.
+     */
+    void to_dimacs(std::ostream& os) const;
+
+    /**
+     * Activates the PBlib to generate smaller instance sizes.
+     *
+     * Value will be used during the next \ref add_q_cylinder call.
+     */
+    bool use_pblib = false;
```
[the remark this answers is not in the document]
Comment (cernoch) [reply src/main/cpp/cutcount.h]: Well, that's certainly an option.

I used to set these parameters in the constructor, but gradually, as the number of parameters grows and the number of constructors grows, the amount of boilerplate code grows quadratically. :slightly_frowning_face: 

The `mks::plexator` has 7 parameters and 2 constructors...
https://github.com/master-keying/mks/blob/1b24ecf702d6f6f10674ca88ea3cf57dbec921eb/src/main/cpp/plexator.h#L284-L412

Moreover, from a usage point-of-view so far, it's been fine to 1) call constructor 2) set parameters, 3) start using the class.

File: src/main/cpp/cutcount.h [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+     * Propagate all unit clauses in \ref m_formula.
+     *
+     * This might help a few buggy #SAT solvers (such as #SAT).
+     * However, not every sharpSAT bug is resolved by this trick.
+     */
+    void propagate();
+
+protected:
+
+    //! Geomentry that defines the shape
```
[the remark this answers is not in the document]
Comment (cernoch) [reply src/main/cpp/cutcount.h]: Fixed.

File: src/main/cpp/cutcount.cpp [hunk 8]
```diff
@@ -0,0 +1,331 @@
...
+        // A) at least 1 depth is active
+        clause at_least_one;
+        for (depth_id d = 0; d < geom.depth_max(); ++d) {
+            auto lit = sharpSAT::LiteralID(var(p,d), true);
+            at_least_one.push_back(lit);
+        }
+        m_formula.push_back(std::move(at_least_one));
+
+        // B) at most 1 depth is active
+        for (depth_id d1 = 0; d1 < geom.depth_max() - 1; ++d1) {
```
[the remark this answers is not in the document]
Comment (cernoch) [reply src/main/cpp/cutcount.cpp]: Will investigate this before merging the PR.

File: src/test/cpp/cutcount_test.cpp [hunk 9]
```diff
@@ -0,0 +1,745 @@
...
+        counter.add_within_cylinder({{0},{0,1},{1}});
+        REQUIRE(counter.count_cuttings() == 2);
+    }
+
+    SECTION("overlap of two cylinders") {
+        counter.add_within_cylinder({{0,1},{0,1},{1}});
+        //REQUIRE(counter.count_cuttings() == 4);
+        counter.add_within_cylinder({{0},{0,1},{0,1}});
+
+        std::ofstream ofs("overlap.cnf");
```
Comment (horenmar) [inline src/test/cpp/cutcount_test.cpp]: This seems to be a debugging left-over (also there is an assertion commented-out above it).

**Comment by horenmar** [comment]
Interestingly, dumping the CNF to dimacs and then feeding that to the sharpSAT frontend works without problems. This suggests that there is something wrong with how we use the API.

**Comment by horenmar** [comment]
I added a temporary hotfix for this issue to sharpSAT, but it should be investigated anyway.

**Comment by horenmar** [comment]
We will need to start splitting apart tests by their runtime.

The new tests for `cutcount` take well over 30s in _release_ mode.

**Comment by horenmar** [comment]
What state is this currently in? I see the commits are squashed and f30151d2639f643bb606f9a4d8c2c26024c4089b attempts to address #49. Did it work, or should I keep working on it?

**Comment by cernoch** [comment]
I've decided not to merge this yet, because the [integration tests are failing](https://ci.appveyor.com/project/horenmar/mks/build/728/job/72955851xyrp815k):

>     Start 29: cutcount-cli-zetta
>     28/33 MemCheck #29: cutcount-cli-zetta ...............***Failed  Required regular expression not found.Regex=[^1000000000000000000000
>     $
>     ]  1.04 sec
>     === Marker ===: Cmd: OpenCppCoverage --export_type binary:cov-report29.bin --quiet --sources C:\projects\mks --cover_children --  C:/projects/mks/Build/Debug/mks-cutting-counter.exe C:/projects/mks/inputs/geometries/zetta_geometry.txt
>     1000000000000000000000

Apparently, there is some issue with the `coverage-helper`. Even though the binary _does_ print `1000000000000000000000\n`, ctest does not recognize it. I've tried capturing `1000000000000000000000\r\n`, still without success.

**Comment by cernoch** [comment]
My current plan:

1) Rewrite `coverage-helper` to PowerShell (see my comment in #49)
2) Make sure the cmake-regex issue is fine. This should also fix #49.
3) Re-run the tests and merge the PR.

**Comment by horenmar** [comment]
Have we compared the performance of simple constraints using PBLib vs without yet? I remember we talked about whether it is worth it to use it in the constructor.

**Comment by horenmar** [comment]
I also think the "exhaustive fuzzer" test should go into its separate `TEST_CASE` and we should introduce a new tag, like `[.long]`, that hides long-running tests from normal testing (and move them into integration testing).

Otherwise LGTM and we can finaly merge this.

**Pull request merged**